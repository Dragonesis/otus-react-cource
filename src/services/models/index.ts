export interface Unit {
  id: string
  name: string
  power: number
  defense: number
  className: 'healer' | 'magician' | 'warrior'
}

export type DeckOfCard = {
  id: string
  name: string
  cards: Unit[]
}

export type DeckOfCardList = DeckOfCard[]

export interface User {
  name: string
  email: string
  birthday: string
}
