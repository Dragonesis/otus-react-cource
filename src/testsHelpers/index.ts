export { customRender } from './customRender'
export { checkingCardAttributes } from './checkingCardAttributes'
