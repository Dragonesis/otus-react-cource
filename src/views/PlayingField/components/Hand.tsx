import React, { FC } from 'react'
import styled from '@emotion/styled'
import { useStore } from '@/services/adapters/store'
import { Card } from './Card'

export const Hand: FC = () => {
  const { cardsInHand, setCardsInHand } = useStore()

  return (
    <Core>
      {cardsInHand?.map((unit) => (
        <Card
          key={unit.id}
          unit={unit}
          handleClick={() => setCardsInHand(cardsInHand.filter(({ id }) => id !== unit.id))}
        />
      ))}
    </Core>
  )
}

const Core = styled.div`
  display: flex;
`
