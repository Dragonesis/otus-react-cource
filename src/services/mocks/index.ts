import { DeckOfCardList } from '../models'

export const deckOfCardList: DeckOfCardList = [
  {
    id: '06f9c59c-6789-4474-bea7-1c8852b95d99',
    name: 'Nichole',
    cards: [
      {
        id: '1b1069b8-ccf2-4218-a587-5d8e3dce3f74',
        name: 'Pearl',
        power: 7,
        defense: 3,
        className: 'magician',
      },
      {
        id: 'c5f8d91e-ecc2-45e8-b334-e02e1c255077',
        name: 'Marylou',
        power: 3,
        defense: 8,
        className: 'warrior',
      },
      {
        id: 'ed01b0fb-60b6-40a1-8374-ae483471ebc1',
        name: 'Helen',
        power: 2,
        defense: 2,
        className: 'healer',
      },
      {
        id: '499acfdd-9f59-463d-bf2c-3479a07df837',
        name: 'Mae',
        power: 8,
        defense: 4,
        className: 'warrior',
      },
      {
        id: '92915c1e-d868-4fd5-9114-f973ca94cf6b',
        name: 'Kirkland',
        power: 6,
        defense: 2,
        className: 'magician',
      },
      {
        id: '1d5f48d0-c594-4ff7-8a09-56850b5b2ee7',
        name: 'Sheri',
        power: 6,
        defense: 0,
        className: 'magician',
      },
      {
        id: 'fb3dc740-f271-4802-b4d6-a4e21dfb1b2d',
        name: 'Spence',
        power: 5,
        defense: 0,
        className: 'warrior',
      },
      {
        id: '15b958c3-fc94-4a69-a226-67a65e810888',
        name: 'Susan',
        power: 8,
        defense: 5,
        className: 'magician',
      },
      {
        id: '6bb42e22-9264-4205-9bd3-add6f49a4a14',
        name: 'Ginger',
        power: 1,
        defense: 5,
        className: 'magician',
      },
      {
        id: '1c732899-558f-4025-a84d-43f50d2fad52',
        name: 'Gibbs',
        power: 5,
        defense: 10,
        className: 'healer',
      },
      {
        id: 'd5861966-71d7-440f-88b3-20fc0b13a795',
        name: 'Marla',
        power: 10,
        defense: 5,
        className: 'warrior',
      },
      {
        id: 'c55ef6af-7720-439b-8c9b-314ea0140cd0',
        name: 'Ray',
        power: 9,
        defense: 8,
        className: 'warrior',
      },
    ],
  },
  {
    id: '4a5d299c-1b8b-434d-8249-a2d7fb224a3d',
    name: 'Caitlin',
    cards: [
      {
        id: '83ee4db2-d056-41c0-9297-a53ca57a923a',
        name: 'Campbell',
        power: 4,
        defense: 7,
        className: 'magician',
      },
      {
        id: '82610421-0a71-4e2d-b6af-c52f5f34d823',
        name: 'Araceli',
        power: 2,
        defense: 8,
        className: 'healer',
      },
      {
        id: '8f864d59-166b-4f86-9014-130b2c9ade2d',
        name: 'Jill',
        power: 7,
        defense: 5,
        className: 'magician',
      },
      {
        id: '7c17edd6-530a-4c11-9f90-97c0f52ef49f',
        name: 'Leanna',
        power: 5,
        defense: 8,
        className: 'warrior',
      },
      {
        id: '42839e99-e807-4cdb-b9ae-34d60486d8a7',
        name: 'Earline',
        power: 5,
        defense: 7,
        className: 'healer',
      },
      {
        id: '13d6fafc-f6af-4239-8317-96eb565f7d9c',
        name: 'Marietta',
        power: 7,
        defense: 7,
        className: 'healer',
      },
      {
        id: '368f97a8-6662-4987-8c12-12b23b184762',
        name: 'Adrienne',
        power: 10,
        defense: 4,
        className: 'warrior',
      },
      {
        id: '56d25945-511f-4c3b-ada3-56f1d84067ac',
        name: 'Thomas',
        power: 6,
        defense: 1,
        className: 'warrior',
      },
      {
        id: '840cfd22-1d32-4a0b-9981-13b48b6c4af6',
        name: 'Olson',
        power: 5,
        defense: 2,
        className: 'magician',
      },
      {
        id: '8f347180-c30a-4dac-a939-d79a297be66e',
        name: 'Wolfe',
        power: 0,
        defense: 1,
        className: 'magician',
      },
      {
        id: 'e376d945-97b1-46c2-a0ee-b1b3325da0ac',
        name: 'Richardson',
        power: 6,
        defense: 1,
        className: 'magician',
      },
      {
        id: 'de2021cf-3e01-4448-bc75-f16f71afedd2',
        name: 'Tammi',
        power: 5,
        defense: 4,
        className: 'magician',
      },
    ],
  },
  {
    id: '40c124e7-a06b-449f-a867-59ce52bf1ad7',
    name: 'Rebecca',
    cards: [
      {
        id: '42f9b57e-0c5a-4569-ad7b-bd9906ba7ee4',
        name: 'Morton',
        power: 4,
        defense: 9,
        className: 'magician',
      },
      {
        id: 'af415f44-b703-4a9a-9a50-4ba292a4ed01',
        name: 'Dianne',
        power: 5,
        defense: 5,
        className: 'magician',
      },
      {
        id: '146ded1d-3e2d-41f7-897d-a6992de220b2',
        name: 'Marina',
        power: 8,
        defense: 5,
        className: 'healer',
      },
      {
        id: 'e3258643-e8a8-44ab-8b03-dcb72fa71120',
        name: 'Ana',
        power: 2,
        defense: 3,
        className: 'healer',
      },
      {
        id: '5da0d207-97cb-48c1-be57-bd3655a137f5',
        name: 'Daisy',
        power: 5,
        defense: 7,
        className: 'warrior',
      },
      {
        id: '58beda7e-f7bf-46bc-9a80-72d7fca2eb84',
        name: 'Mccarty',
        power: 6,
        defense: 9,
        className: 'magician',
      },
      {
        id: 'd894c559-3fb9-4aae-b685-afee1effe829',
        name: 'Burnett',
        power: 3,
        defense: 0,
        className: 'magician',
      },
      {
        id: '36bcaa84-3b4d-41ba-87b4-2fb255472a34',
        name: 'Alvarado',
        power: 5,
        defense: 7,
        className: 'healer',
      },
      {
        id: '680379d5-d33a-4cb0-89bc-53ac6a67ce79',
        name: 'Ellis',
        power: 2,
        defense: 8,
        className: 'healer',
      },
      {
        id: 'ffd83c8b-7a90-4861-b0d2-796767929969',
        name: 'Cristina',
        power: 2,
        defense: 4,
        className: 'healer',
      },
      {
        id: '1d7f5229-a42d-4901-aadb-6f25d40888d8',
        name: 'Allen',
        power: 4,
        defense: 6,
        className: 'magician',
      },
      {
        id: '6cbcb341-f2ea-4cf7-9f6f-3a38f9d7e166',
        name: 'Davenport',
        power: 3,
        defense: 3,
        className: 'healer',
      },
    ],
  },
  {
    id: '3a3c6516-bf87-428f-b523-8ff5c58fa087',
    name: 'Swanson',
    cards: [
      {
        id: '7856fb17-566d-40c6-8a73-1a88a56bf922',
        name: 'Leslie',
        power: 1,
        defense: 3,
        className: 'warrior',
      },
      {
        id: 'e2c370b1-a2c1-481c-927a-a146b80557e1',
        name: 'Barlow',
        power: 7,
        defense: 6,
        className: 'healer',
      },
      {
        id: '7c4f9a47-7da3-40b8-b949-2ff972f7b01d',
        name: 'Joyner',
        power: 4,
        defense: 8,
        className: 'magician',
      },
      {
        id: 'a520a435-5013-40e1-ab15-1e9b08a015f4',
        name: 'Henry',
        power: 10,
        defense: 7,
        className: 'healer',
      },
      {
        id: 'd8c4a2d5-bcd9-4427-bb46-a4417e9d33c1',
        name: 'Moon',
        power: 2,
        defense: 9,
        className: 'warrior',
      },
      {
        id: '0e3b781f-e0dd-4e73-9c09-341c4c044ca2',
        name: 'Leann',
        power: 4,
        defense: 6,
        className: 'healer',
      },
      {
        id: '8cd17ae6-99a7-49e4-86c4-0688ba30dc55',
        name: 'Simon',
        power: 6,
        defense: 3,
        className: 'healer',
      },
      {
        id: '4fc14c46-c050-43fc-8807-531446986ca3',
        name: 'Amie',
        power: 3,
        defense: 7,
        className: 'magician',
      },
      {
        id: '03778717-757e-4716-bcac-72e4ac8ec03f',
        name: 'Kara',
        power: 9,
        defense: 10,
        className: 'magician',
      },
      {
        id: '67fd1a49-36e1-4da5-9c96-0595889706e0',
        name: 'Gina',
        power: 7,
        defense: 9,
        className: 'magician',
      },
      {
        id: 'ae29204d-63dd-46b1-96d3-f0d763231473',
        name: 'Lindsey',
        power: 5,
        defense: 6,
        className: 'warrior',
      },
      {
        id: '00037bf3-05de-495c-b8f8-661ec569f5e4',
        name: 'Burgess',
        power: 0,
        defense: 7,
        className: 'warrior',
      },
    ],
  },
]
