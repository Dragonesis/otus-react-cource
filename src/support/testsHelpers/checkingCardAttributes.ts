import { screen } from '@testing-library/react'
import { Unit } from '@/services/models'

export const checkingCardAttributes = (unit: Unit) => {
  const card = screen.queryAllByLabelText('Карта ' + unit.name)[0]
  expect(card).toBeInTheDocument()
  expect(screen.getByText(unit.name)).toBeInTheDocument()
  expect(screen.getByText('Сила: ' + unit.power)).toBeInTheDocument()
  expect(screen.getByText('Защита: ' + unit.defense)).toBeInTheDocument()
  expect(screen.getByText('Класс: ' + unit.className)).toBeInTheDocument()
}
