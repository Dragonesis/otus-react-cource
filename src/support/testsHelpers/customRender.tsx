import React from 'react'
import { render } from '@testing-library/react'
import { StoreContext } from '@/services/adapters/store'
import { DeckOfCard, Unit } from '@/services/models'

export const customRender = (
  children: React.ReactNode,
  {
    providerProps,
    ...renderOptions
  }: {
    providerProps: {
      cardDeck: DeckOfCard | undefined
      cardsInHand: Unit[]
      setDeckOfCard: (arg: DeckOfCard) => DeckOfCard
      setCardsInHand: (arg: Unit[]) => Unit[]
    }
  },
) => {
  return render(<StoreContext.Provider value={providerProps}>{children}</StoreContext.Provider>, renderOptions)
}
