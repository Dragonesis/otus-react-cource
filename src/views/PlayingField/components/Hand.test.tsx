import React from 'react'
import { render, screen } from '@testing-library/react'
import { deckOfCardList } from '@/services/mocks'
import { customRender, checkingCardAttributes } from '@/support/testsHelpers'

import { Hand } from './Hand'

test('Hand renders correctly', () => {
  const { asFragment } = render(<Hand />)
  expect(asFragment()).toMatchSnapshot()
})

describe('Hand with cards available', () => {
  const setDeckOfCard = jest.fn()
  const setCardsInHand = jest.fn()

  const cards = deckOfCardList[0].cards

  const providerProps = {
    cardDeck: undefined,
    cardsInHand: [cards[0], cards[1], cards[2]],
    setDeckOfCard,
    setCardsInHand,
  }

  customRender(<Hand />, { providerProps })
  const cardsOnScreen = screen.queryAllByLabelText(/Карта/i)
  expect(cardsOnScreen).toHaveLength(3)

  it.each([[cards[0]], [cards[1]], [cards[2]]])('%s', (unit) => {
    customRender(<Hand />, { providerProps })
    checkingCardAttributes(unit)
  })
})
