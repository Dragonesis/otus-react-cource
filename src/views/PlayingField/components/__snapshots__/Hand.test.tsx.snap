// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hand renders correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<div
    class="emotion-0"
  />
</DocumentFragment>
`;
